Bio/SeqIO/SnapGeneIO.py

```diff
@@ -305,10 +331,14 @@ def iterate(self, handle):
             raise ValueError("The file does not start with a SnapGene cookie packet")
         _parse_cookie_packet(length, data, record)
 
+        helper = XmlHelper()
         for (packet_type, length, data) in packets:
             handler = _packet_handlers.get(packet_type)
             if handler is not None:
-                handler(length, data, record)
+                handler(length, data, record, helper)
+
+        if helper.transformed:
+            warnings.warn("Some text values have been transformed", BiopythonWarning)
```

Comment:
Warnings are dangerous; they get emitted only once in a Python session.

Reply:
We already have multiple parser warnings in the GenBank parser and elsewhere in Biopython, but @mdehoon and I disagree on the use of warnings. This came up on another issue recently too.

I am of the view that https://docs.python.org/3/library/warnings.html provides the tools needed for a user to decide how to deal with warnings as they wish (including silencing them all, showing them all, or upgrading them to errors), and we just need to word the warning appropriately to balance their usefulness by default.

e.g. If a warning might be triggered dozens of times on a file, use the same warning string so that by default it is only seen once.

In this case *if* we want to have lots of warnings shown by default, you could issue a warning on each transformation using problematic value.

Reply:
@gouttegd  One solution is to use a `strict=True` argument to the `__init__` function of `SnapGeneIterator` that specifies if the parser should raise an exception (if `strict` is `True`) or ignore it (if `strict` is `False`). And then we need to convince @peterjc  that this is a better solution than using the `warnings` module, and allow for optional arguments to `Bio.SeqIO.parse`.

The `parse` function in `Bio.motifs` uses this approach.

Reply:
@peterjc  Few users will know the `warnings` module in sufficient detail to be able to do this. Also, the `warnings` module allows for global filtering of warnings only, whereas in a typical use case, a user may want to ignore this warning about SnapGene, but continue to see warnings from other Biopython functions or e.g. numpy.

Reply:
I am extremely reluctant to raise an exception in this case, even with a hypothetical `strict=True` option.

The presence of newline characters in a qualifier value, in a SnapGene file, is not an error. The SnapGene format does allow that. A SnapGene file with newline characters in some qualifiers is still a valid SnapGene file.

The reason we want to replace those characters when parsing is not because they should not be there (again, the format allows them to be there), but because most downstream code working with `SeqFeature` objects (such as the GenBank writer) would not expect such newline characters.

Raising an exception to say “we won’t parse this file any further, it contains invalid characters” would be flatly wrong IMHO.

Reply:
Besides, what should the default value of the `strict=` option be?

If it is `True`, then by default we would refuse to parse perfectly valid files. Not great.

If it is `False`, then by default the user would have _no way at all_ to known that a data transformation occurred, until they parse the file again with `strict=True` (why would they do that?). I fail to see how this is an improvement compared to the use of warnings, which at least have a chance of being seen.

Reply:
@gouttegd  If the SnapGene format allows newline characters, then the SnapGene parser should neither raise an exception nor emit a warning. Instead, the Genbank writer should raise an Exception (or replace newlines by spaces) if it finds a newline character.

Reply:
Accepting new lines in the SnapGene files (and raising a ValueError in the GenBank writer) would be reasonable. My impression from the issue report is this is rare and likely a glitch, so I would be happy with the warning (as in this PR), or forcing the user to correct the annotation before writing it.

And to Michiel, if you wanted to silence or otherwise specially treat warnings from ``SeqIO.parse(...)`` you could use a context manager https://docs.python.org/3/library/warnings.html#temporarily-suppressing-warnings (as we do in our unit tests for verifying a warning is raised). We could add this to our documentation, ideally with a good motivational use case?

Reply:
> And to Michiel, if you wanted to silence or otherwise specially treat warnings from SeqIO.parse(...) you could use a context manager https://docs.python.org/3/library/warnings.html#temporarily-suppressing-warnings (as we do in our unit tests for verifying a warning is raised). We could add this to our documentation, ideally with a good motivational use case?

It's too complicated. Only advanced users will know about this. Not to mention that warnings are easy to miss if a script prints lots of output to the screen.
Raising a `ValueError`, on the other hand, can simply tell the user to use `strict=False` if they are sure they want to ignore the issue found.

Reply:
@gouttegd  Btw from your example it looks like the line break in the SnapGene file was just included for formatting and does not convey any information. Then you could just let the SnapGene parser always replace the line break with a space, without a warning, and mention somewhere in the documentation that the SnapGene parser does this.
`textwrap.shorten(yourtext, width=sys.maxsize)` lets you do this in a clean way.

Reply:
> My impression from the issue report is this is rare and likely a glitch

I have never seen it myself in the SnapGene files I’ve got, but in the case reported in #3885 the newline is represented in the file as a XML entity `&x10;`. I doubt this can happen as a result of a glitch.

> forcing the user to correct the annotation before writing it

I am very much against that.

Because the SnapGene format is binary, practically speaking SnapGene files can only be edited using the SnapGene software. The reason I wrote the SnapGene parser in the first place was so that I could deal with SnapGene files (that I had received from collaborators who didn’t bother to use a more standard format) without having to install SnapGene, that I personally do not use.


> Then you could just let the SnapGene parser always replace the line break with a space, without a warning, and mention somewhere in the documentation that the SnapGene parser does this.

I’m fine with advertising the line-break removal in the documentation, but I don’t see why the warning should be removed. It doesn’t hurt and it gives the user (who may not have read the documentation) one more chance to be aware of the removal. What is the problem with that?

Reply:
Let's think about this a bit more carefully. Is our basic assumption that newlines in SnapGene files are
(a) informative;
(b) not informative (i.e. they were just added for formatting, but do not carry information);
(c) sometimes informative and sometimes not informative?